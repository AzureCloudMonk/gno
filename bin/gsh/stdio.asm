**************************************************************************
*
* The GNO Shell Project
*
* Developed by:
*   Jawaid Bazyar
*   Tim Meekins
*
* $Id: stdio.asm,v 1.5 1998/09/08 16:53:14 tribby Exp $
*
**************************************************************************
*
* STDIO.ASM
*   By Tim Meekins
*   Modified by Dave Tribby for GNO 2.0.6
*
* This is the custom stdio for the shell
*
* stdout buffer size: 512 bytes
* stderr buffer size: 256 bytes
* stdin  buffer size: 128 bytes
*
* Note: text set up for tabs at col 16, 22, 41, 49, 57, 65
*              |     |                  |       |       |       |
*	^	^	^	^	^	^	
**************************************************************************

	mcopy /obj/gno/bin/gsh/stdio.mac

dummystdio	start		; ends up in .root
	end

	setcom 60

**************************************************************************
*
* putchar - output a character to standard out
* On entry: A = character
*
**************************************************************************

putchar	START

	using	stdout
	
	tay		Note: lock destroys Acc
	lock	mutex	Wait for others to leave, and lock.
	tya
	and	#$FF	Isolate the single character.
	ldx	index	Get num of chars already in stream.
	sta	stream,x	Store this char + null byte.
	inx		Bump length of stream by one.
	cmp	#13	If character was newline,
	beq	_flush	 go write the stream.
	cpx	#512	If length < 512,
	bcc	done	 all done.
_flush	stx	index	Save current length.
	Write	WriteParm	Write the stream.
	ldx	#0	Set new length to 0.

done	stx	index	Save stream length in global.
	unlock mutex	Allow others through.
	rts		Return to caller.

	END

**************************************************************************
*
* puts - output a c string to standard out
* On entry: A/X = pointer to string
*
**************************************************************************

puts	START

	using	stdout

	tay		Note: lock destroys Acc
	lock	mutex	Wait for others to leave, and lock.
	sty	getchar+1	Save low-order bytes of address.
	txa

	short	a	SWITCH TO SINGLE-BYTE MEMORY MODE.
	sta	getchar+3	Store high-order byte of address.
	ora	getchar+2	If string address
	ora	getchar+1	 is 00/0000,
	beq	exit	  don't do the write.

	ldy	index	Get current number of chars in stream.
	ldx	#0	Clear source string offset.

getchar	lda	>$FFFFFF,x	Get next character from string.
	beq	done	Done when we see a null byte.
	sta	stream,y	Store in output stream.
	iny		Bump the stream and
	inx		 string pointers.
	cmp	#13	If newline was encountered,
	beq	_flush	 go write & flush the stream.
	cpy	#512	If stream length < 512,
	bcc	getchar	 continue copying characters.

_flush	sty	index	Save length of stream.
	phx		Hold source string offset on stack.
	long	a	SWITCH TO FULL-WORD MEMORY MODE.
	Write	WriteParm	Write the stream to stdout
	Flush	flushparm	 and flush it.
	short	a	SWITCH TO SINGLE-BYTE MEMORY MODE.
	plx		Restore source string offset to X-reg.
	ldy	#0	Set stream length to 0.
	bra	getchar	Continue copying characters.

; Arrive here when null character is encountered.
done	sty	index	Save stream length in global.

exit	long	a	SWITCH TO FULL-WORD MEMORY MODE.
	unlock mutex	Allow others through.
	rts		Return to caller.

	END	              


**************************************************************************
*
* flush - flush stdout
*
**************************************************************************

flush	START

	using	stdout
	
	lock	mutex
	lda	index
	beq	skip
	Write	WriteParm
	Flush	flushparm
	stz	index            
skip	unlock mutex

	rts	

	END

**************************************************************************
*
* data for standard out
*
**************************************************************************

stdout	PRIVDATA

mutex	key

WriteParm	dc	i2'4'
	dc	i2'2'	;2 is standard out
	dc	i4'stream'
index	dc	i4'0'
	dc	i4'0'

flushparm	dc	i2'1'
	dc	i2'2'

stream	ds	512+1

	END

**************************************************************************
*
* errputchar - output a character to standard error
* On entry: A = character
*
**************************************************************************

errputchar	START

	using	stderr
	
	tay		;lock destroys Acc
	lock	errmutex
	tya
	and	#$FF
	ldx	errindex
	sta	errstream,x
	inx
	cmp	#13
	beq	_flush
	cpx	#256
	bcc	done
_flush	stx	errindex
	Write	errWriteParm
	ldx	#0
done	stx	errindex
	unlock errmutex
	rts

	END

**************************************************************************
*
* errputs - output a c string to standard error
* On entry: A/X = pointer to string
*
**************************************************************************

errputs	START

	using	stderr

	tay		;lock destroys Acc
	lock	errmutex
	sty	getchar+1
	txa

	short	a
	sta	getchar+3

	ldy	errindex
	ldx	#0

getchar	lda	>$FFFFFF,x
	beq	done
	sta	errstream,y
	iny
	inx
	cmp	#13
	beq	_flush
	cpy	#256
	bcc	getchar

_flush	sty	errindex
	phx
	long	a
	Write	errWriteParm
	short	a
	plx
	ldy	#0
	bra	getchar

done	sty	errindex
	long	a
	unlock errmutex
	rts

	END	              

**************************************************************************
*
* errflush - flush stderr
*
**************************************************************************

errflush	START

	using	stderr
	
	lock	errmutex
	Write	errWriteParm
	stz	errindex	 
	unlock errmutex

	rts	

	END

**************************************************************************
*
* data for standard error
*
**************************************************************************

stderr	PRIVDATA

errmutex	key

errWriteParm	dc	i2'4'
	dc	i2'3'	;3 is standard err
	dc	i4'errstream'
errindex	dc	i4'0'
	dc	i4'0'

errstream	ds	256+1	;not as large as stdout

	END

**************************************************************************
*
* getchar - read a single character from standard input
* on exit: a = character, -1 if EOF
*
**************************************************************************

getchar	START

	using	stdin

	lock	inmutex

	lda	insize	;any characters in stream?
	bne	grabchar	;yup

readloop	Read	inReadParm
	bcc	okread
	ldy	#-1	;return EOF on ALL errors
	jmp	done2

okread	stz	inindex
	lda	insize
	bne	grabchar

	cop	$7F	;no characters ready, so wait
	bra	readloop

grabchar	ldx	inindex
	lda	instream,x
	and	#$7F
	inc	inindex
	dec	insize

	tay

done2	unlock inmutex
	tya
	rts

	END

**************************************************************************
*
* data for standard input
*
**************************************************************************

stdin	PRIVDATA

inmutex	key

inReadParm	dc	i2'4'
	dc	i2'1'	;1 is standard input
	dc	i4'instream'
inrequest	dc	i4'128'
insize	dc	i4'0'

inindex	dc	i2'0'

instream	ds	128+1

	END
