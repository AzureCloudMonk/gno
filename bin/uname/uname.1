.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)uname.1	8.3 (Berkeley) 4/8/94
.\"
.\" Modified for GNO (Apple IIgs) by Steve Reeves and Devin Reade's bsd2man
.\"
.\" $Id: uname.1,v 1.1 1998/02/17 02:50:28 gdr-ftp Exp $
.\"
.TH UNAME 1 "January 23, 1998" GNO "Commands and Applications"
.SH NAME
.B uname
\- display information about the system
.SH SYNOPSIS
.B uname
[
.B \-amnrsv
]
.br
.B uname
.B \-V
.SH DESCRIPTION
By default, the
.B uname
command writes the name of the operating system implementation to
standard output.
When options are specified, strings representing one or more system
characteristics are written to standard output.
.SH OPTIONS
.IP \fB\-a\fP
Behave as though the options
.BR \-m ,
.BR \-n ,
.BR \-r ,
.BR \-s ,
and
.B \-v
were specified.
.IP \fB\-m\fP
Write the type of the current hardware platform to standard output.
.IP  \fB\-n\fP
Write the name of the system to standard output.
.IP \fB\-r\fP
Write the current release level of the operating system
to standard output.
.IP \fB\-s\fP
Write the name of the operating system implementation to standard output.
.IP \fB\-v\fP
Write the version level of this release of the operating system
to standard output.
.IP \fB\-V\fP
Write version information for the
.B uname
command and exit.
.LP
If the
.B \-a
flag is specified, or multiple flags are specified, all
output is written on a single line, separated by spaces.
.SH "RETURN VALUES"
The
.B uname
utility exits 0 on success, and >0 if an error occurs.
.SH VERSION
This manual page documents
.B uname
version 3.0.
.SH HISTORY
Version 1.0 (February 1994) of
.B uname
for GNO was written by Phil Vandry <vandry@cam.org>.
Version 2.0 (September 1994)
was written by Ian Schmidt <irsman@drift.winternet.com>.
.SH STANDARDS
The
.B uname
command is expected to conform to the
.SM IEEE
Std1003.2 (``POSIX.2'')
specification.
.SH SEE ALSO
.BR uname (3)
