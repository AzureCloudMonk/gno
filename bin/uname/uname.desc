Name:    uname
Version: 3.0 (January 23, 1998)
Shell:   GNO
Author:  Steve Reeves (from FreeBSD code)
Contact: stever@gate.net
Where:   /bin
FTP:     ground.isca.uiowa.edu apple2.caltech.edu trenco.myrias.com

   Display information about the system.
