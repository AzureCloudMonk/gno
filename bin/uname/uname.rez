/*
 * Resources for uname
 *
 * $Id: uname.rez,v 1.1 1998/02/17 02:50:32 gdr-ftp Exp $
 */

#define PROG "uname"
#define DESC "Display information about the system."

#include "Types.rez"

/*
 * Version
 */
resource rVersion (1, purgeable3) {
	{ 3, 0, 0,	/* Version 3.0.0 */
	  release,	/* development|alpha|beta|final|release */
	  0 },		/* non-final release number */
	verUS,	/* Country */
	PROG,	/* Program name */
        DESC " Released with GNO/ME."
};


/*
 * Comment
 */
resource rComment (1, purgeable3) {
	PROG " v3.0 (January 23, 1998)\n"
	"GNO utility: " DESC "\n"
	"Ported from FreeBSD code by Steve Reeves."
};
