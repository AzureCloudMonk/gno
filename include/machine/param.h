/*
 * machine/param.h
 *
 * Copyright 1991-1996, Procyon, Inc.
 * All Rights Reserved
 *
 * $Id: param.h,v 1.1 1997/02/28 04:42:07 gdr Exp $
 */

#ifndef _MACHINE_PARAM_H_
#define _MACHINE_PARAM_H_

#define MACHINE "APPLEIIGS"

#ifdef KERNEL			/* non-BSD.  Obsolete? */
#define DAYSPERNYEAR 365
#define SECSPERDAY 86400L
#endif

/*
 * Round p (pointer or byte index) up to a correctly-aligned value
 * for all data types (int, long, ...).   The result is unsigned int
 * and must be cast to any desired pointer type.
 */
#define ALIGNBYTES	0
#define ALIGN(p)	p

#endif
