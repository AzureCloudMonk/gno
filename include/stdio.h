/****************************************************************
*
*  stdio.h - input/output facilities
*
*  February 1989
*  Mike Westerfield
*
*  Copyright 1989, 1993
*  Byte Works, Inc.
*
****************************************************************/

#ifndef __stdio__
#define __stdio__

#ifndef _SYS_CDEFS_H_
#include <sys/cdefs.h>
#endif

#ifndef _SYS_SYSLIMITS_H_
#include <sys/syslimits.h>
#endif

/*
 *  Misc.
 */

#ifndef __va_list__
#define __va_list__
typedef char *__va_list[2];
#endif

#ifndef EOF
#define EOF             (-1)
#endif

#ifndef NULL
#define NULL  (void *) 0L
#endif

#ifndef __size_t__
#define __size_t__ 1
typedef unsigned long size_t;
#endif

/* seek codes */

#define SEEK_SET        0
#define SEEK_CUR        1
#define SEEK_END        2

/*
 *  Error handling
 */

#if !defined(__KeepNamespacePure__) && !defined(_ANSI_SOURCE) && !defined(_POSIX_SOURCE)
   extern const int sys_nerr;               /* largest index for sys_errlist */
   extern const char * const sys_errlist[]; /* error messages */
#endif

/*
 *  files
 */

typedef struct __file {
   struct __file        *next;          /* next file in linked list */
   unsigned char        *_ptr,          /* next location to write to */
                        *_base,         /* first byte of the buffer */
                        *_end;          /* end of the file buffer */
   unsigned long        _size,          /* size of the file buffer */
                        _cnt;           /* # chars that can be read/writen to buffer */
#if defined(__ORCAC_VERSION) && (__ORCAC_VERSION > 210)
   int                  _pbk[2];        /* put back character; changed in v2.1.1b2 */
#else
#error bad stuff batman
   int                  _pbk;           /* put back character */
#endif                        
   unsigned int         _flag,          /* buffer flags */
                        _file;          /* GS/OS file ID */
   } FILE;

#define BUFSIZ          1024            /* default buffer size */
#define _LBUFSIZ        255             /* line buffer size */

#define _IOFBF          0x0001          /* full buffering */
#define _IONBF          0x0002          /* no buffering */
#define _IOLBF          0x0004          /* flush when a \n is written */
#define _IOREAD         0x0008          /* currently reading */
#define _IOWRT          0x0010          /* currently writing */
#define _IORW           0x0020          /* read/write enabled */
#define _IOMYBUF        0x0040          /* buffer was allocated by stdio */
#define _IOEOF          0x0080          /* has an EOF been found? */
#define _IOERR          0x0100          /* has an error occurred? */
#define _IOTEXT         0x0200          /* is this file a text file? */
#define _IOTEMPFILE     0x0400          /* was this file created by tmpfile()? */

extern FILE *stderr;                    /* standard I/O files */
extern FILE *stdin;
extern FILE *stdout;

#define L_tmpnam        26              /* size of a temp name */
#define TMP_MAX         10000           /* # of unique temp names */
#ifdef __GNO__
#define	__max_open	OPEN_MAX	/* <= OPEN_MAX <sys/syslimits.h> */
#else
#define __max_open	32767		/* Orca Shell: no practical limit */
#endif
#ifndef __KeepNamespacePure__
   #define SYS_OPEN	__max_open      /* max # open files */
#endif
#define FOPEN_MAX       __max_open      /* max # open files */
#undef	__max_open
#define FILENAME_MAX    1024            /* recommended file name length */

/*
 *  Other types
 */

typedef long fpos_t;

/*
 *  Functions declared as macros
 */

#ifdef __ORCAC_VERSION		/* Orca/C v2.1.0 or later */

#define setbuf(stream,buf)      ((buf==NULL) ? (void) __setvbuf(stream,NULL,_IONBF,0l) : (void) __setvbuf(stream,buf,_IOFBF,(size_t) BUFSIZ))
#define setlinebuf(stream)	(__setvbuf((stream),NULL,_IOLBF,_LBUFSIZ))
#define rewind(stream)          (__fseek((stream),0L,SEEK_SET))
int		__fseek(FILE *, long, int);
int		__setvbuf(FILE *, char *, int, size_t);
int             getc(FILE *);
int             putc(int, FILE *);

#else	/* __ORCAC_VERSION */

#define getc(p)                 (fgetc(p))
#define putc(x, p)              fputc(x, p)
#define setbuf(stream,buf)      ((buf==NULL) ? (void) setvbuf(stream,NULL,_IONBF,0l) : (void) setvbuf(stream,buf,_IOFBF,(size_t) BUFSIZ))
#define setlinebuf(stream)	(setvbuf((stream),NULL,_IOLBF,_LBUFSIZ))
#define rewind(stream)          (fseek((stream),0L,SEEK_SET))

#endif	/* __ORCAC_VERSION */

/*
 *  Function declarations
 */

void            clearerr(FILE *);
int             fclose(FILE *);
int             feof(FILE *);
int             ferror(FILE *);
int             fflush(FILE *);
int             fgetc(FILE *);
int             fgetpos(FILE *, fpos_t *);
char           *fgets(char *, int, FILE *);
FILE           *fopen(const char *, const char *);
int             fprintf(FILE *, const char *, ...);
int             fputc(int, FILE *);
int             fputs(const char *, FILE *);
size_t          fread(void *, size_t, size_t, FILE *);
FILE           *freopen(const char *, const char *, FILE *);
int             fscanf(FILE *, const char *, ...);
int             fseek(FILE *, long, int);
int             fsetpos(FILE *, const fpos_t *);
long int        ftell(FILE *);
size_t          fwrite(const void *, size_t, size_t, FILE *);
int             getchar(void);
char           *gets(char *);
void            perror(const char *);
int             printf(const char *, ...);
int             putchar(int);
int             puts(const char *);
int             remove(const char *);
int             rename(const char *, const char *);
int             scanf(const char *, ...);
int             setvbuf(FILE *, char *, int, size_t);
int             sprintf(char *, const char *, ...);
int             sscanf(const char *, const char *, ...);
FILE           *tmpfile(void);
char           *tmpnam(char *);
int             ungetc(int c, FILE *);
int             vfprintf(FILE *, const char *, __va_list);
int             vprintf(const char *, __va_list);
int             vsprintf(char *, const char *, __va_list);

/*
 * The remaining functions and macros in this file are GNO-specific
 */

/* System V/ANSI C; this is the wrong way to do this, do *not* use these. */
#ifndef _ANSI_SOURCE
#define	P_tmpdir	"/tmp/"
#endif

/*
 * Functions defined in POSIX 1003.1.
 */
#ifndef _ANSI_SOURCE
#define	L_cuserid	9	/* size for cuserid(); UT_NAMESIZE + 1 */
#define	L_ctermid	1024	/* size for ctermid(); PATH_MAX */
#define	fileno(f)	(f->_file)

__BEGIN_DECLS
#ifndef _POSIX_SOURCE
char	*cuserid __P((char *));				/* non-BSD */
#endif
char	*ctermid __P((char *));
FILE	*fdopen __P((int, const char *));
int	 fileno __P((FILE *));
__END_DECLS
#endif /* not ANSI */

/*
 * Routines that are purely local.
 */
#if !defined (_ANSI_SOURCE) && !defined(_POSIX_SOURCE)
__BEGIN_DECLS
char	*fgetln __P((FILE *, size_t *));
int	 fpurge __P((FILE *));
int	 getw __P((FILE *));
int	 pclose __P((FILE *));
FILE	*popen __P((const char *, const char *));
int	 putw __P((int, FILE *));
void	 setbuffer __P((FILE *, char *, int));
char	*tempnam __P((const char *, const char *));
int	 snprintf __P((char *, size_t, const char *, ...));
int	 vfscanf __P((FILE *, const char *, __va_list));
int	 vsnprintf __P((char *, size_t, const char *, __va_list));
int	 vscanf __P((const char *, __va_list));
int	 vsscanf __P((const char *, const char *, __va_list));
FILE	*zopen __P((const char *, const char *, int));
__END_DECLS

#endif	/* !_ANSI_SOURCE && !_POSIX_SOURCE */

#endif	/* __stdio__ */
