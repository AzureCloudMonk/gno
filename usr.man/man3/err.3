.\" Copyright (c) 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	From: @(#)err.3	8.1 (Berkeley) 6/9/93
.\"	$Id: err.3,v 1.3 1998/01/25 20:17:48 gdr-ftp Exp $
.\"
.TH ERR 3 "27 February 1997" GNO "Library Routines"
.SH NAME
.BR err ,
.BR verr ,
.BR errx ,
.BR verrx ,
.BR warn ,
.BR vwarn ,
.BR warnx ,
.BR vwarnx ,
.BR err_set_file ,
.BR err_set_exit
\- formatted error messages
.SH SYNOPSIS
.nf
#include <err.h>

void \fBerr\fR (int \fIeval\fR, const char *\fIfmt\fR, ...);
void \fBverr\fR (int \fIeval\fR, const char *\fIfmt\fR, va_list \fIargs\fR);
void \fBerrx\fR (int \fIeval\fR, const char *\fIfmt\fR, ...);
void \fBverrx\fR (int \fIeval\fR, const char *\fIfmt\fR, va_list \fIargs\fR);
void \fBwarn\fR (const char *\fIfmt\fR, ...);
void \fBvwarn\fR (const char *\fIfmt\fR, va_list \fIargs\fR);
void \fBwarnx\fR (const char *\fIfmt\fR, ...);
void \fBvwarnx\fR (const char *\fIfmt\fR, va_list \fIargs\fR);
void \fBerr_set_file\fR (void *\fIfp\fR);
void \fBerr_set_exit\fR (void (*\fIexitf\fR)(int));
.fi
.SH DESCRIPTION
The
.BR err 
and
.BR warn 
family of functions display a formatted error message on the standard
error output, or on another file specified using the
.BR err_set_file 
function.
In all cases, the last component of the program name, a colon character,
and a space are output.
If the
.IR fmt
argument is not NULL, the formatted error message, a colon character,
and a space are output.
In the case of the
.BR err ,
.BR verr ,
.BR warn ,
and
.BR vwarn 
functions, the error message string affiliated with the current value of
the global variable
.IR errno 
is output.
In all cases, the output is followed by a newline character.
.LP
The
.BR err ,
.BR verr ,
.BR errx ,
and
.BR verrx 
functions do not return, but exit with the value of the argument
.RI ( eval )
The
.BR err_set_exit 
function can be used to specify a function which is called before
.BR exit (3)
to perform any necessary cleanup; passing a null function pointer for
.IR exitf
resets the hook to do nothing.
.SH EXAMPLES
Display the current errno information string and exit:
.RS
.nf

if ((p = malloc(size)) == NULL)
	err(1, NULL);
if ((fd = open(file_name, O_RDONLY, 0)) == -1)
	err(1, "%s", file_name);

.fi
.RE
.LP
Display an error message and exit:
.RS
.nf

if (tm.tm_hour < START_TIME)
	errx(1, "too early, wait until %s", start_time_string);

.fi
.RE
.LP
Warn of an error:
.RS
.nf

if ((fd = open(raw_device, O_RDONLY, 0)) == -1)
	warnx("%s: %s: trying the block device",
	    raw_device, strerror(errno));
if ((fd = open(block_device, O_RDONLY, 0)) == -1)
	err(1, "%s", block_device);

.fi
.RE
.SH "SEE ALSO"
.BR exit (3),
.BR strerror (3)
.SH HISTORY
The
.BR err 
and
.BR warn 
functions first appeared in 4.4BSD.
