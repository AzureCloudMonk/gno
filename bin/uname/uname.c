/*-
 * Copyright (c) 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

/*
 * Modified for GNO (Apple IIGS) by Steve Reeves, January 1998
 *
 * Added conditional compilation based on these macros:
 *
 *   __ORCAC__ for constructs unacceptable to the compiler,
 *   __GNO__ for changes not related to the compiler,
 *   __STDC__ or __P() from <sys/cdefs.h> for ISO C prototyping,
 *   __STACK_CHECK__ for reporting stack space usage,
 *   NOID for excluding copyright and version control strings
 *
 * $Id: uname.c,v 1.1 1998/02/17 02:50:29 gdr-ftp Exp $
 *
 */

#ifndef NOID
#ifndef lint
static char copyright[] =
"@(#) Copyright (c) 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
static char sccsid[] = "@(#)uname.c	8.2 (Berkeley) 5/4/95";
#endif /* not lint */
#endif

#ifndef __GNO__
#include <sys/param.h>
#include <sys/sysctl.h>
#else
#include <sys/utsname.h>
#endif

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

void usage __P((void));

#if defined(__GNO__)  &&  defined(__STACK_CHECK__)
/* Interface to check on how much stack space a C program uses. */
#ifndef _GNO_GNO_H_
#include <gno/gno.h>
#endif
static void report_stack(void)
{
	fprintf(stderr,"\n ==> %d stack bytes used <== \n", _endStackCheck());
}
#endif

int
#ifndef __STDC__
main(argc, argv)
	int argc;
	char *argv[];
#else
main(int argc, char *argv[])
#endif
{
#define	MFLAG	0x01
#define	NFLAG	0x02
#define	RFLAG	0x04
#define	SFLAG	0x08
#define	VFLAG	0x10
	u_int flags;
#ifndef __GNO__
	int ch, mib[2];
	size_t len, tlen;
	char *p, *prefix, buf[1024];
#else
	int ch;
	char *prefix;
	static struct utsname name;
#endif

#if defined(__GNO__)  &&  defined(__STACK_CHECK__)
	_beginStackCheck();
	atexit(report_stack);
#endif

	flags = 0;
	while ((ch = getopt(argc, argv, "amnrsvV")) != -1)
		switch(ch) {
		case 'a':
			flags |= (MFLAG | NFLAG | RFLAG | SFLAG | VFLAG);
			break;
		case 'm':
			flags |= MFLAG;
			break;
		case 'n':
			flags |= NFLAG;
			break;
		case 'r':
			flags |= RFLAG;
			break;
		case 's':
			flags |= SFLAG;
			break;
		case 'v':
			flags |= VFLAG;
			break;
		case 'V':
			printf("uname v3.0\n");
			exit(0);
		case '?':
		default:
			usage();
		}

	argc -= optind;
#ifndef __ORCAC__
	argv += optind;
#else
	argv = argv + optind;
#endif

	if (argc)
		usage();

	if (!flags)
		flags |= SFLAG;

	prefix = "";

#ifndef __GNO__
	if (flags & SFLAG) {
		mib[0] = CTL_KERN;
		mib[1] = KERN_OSTYPE;
		len = sizeof(buf);
		if (sysctl(mib, 2, &buf, &len, NULL, 0) == -1)
			err(1, "sysctl");
		(void)printf("%s%.*s", prefix, len, buf);
		prefix = " ";
	}
	if (flags & NFLAG) {
		mib[0] = CTL_KERN;
		mib[1] = KERN_HOSTNAME;
		len = sizeof(buf);
		if (sysctl(mib, 2, &buf, &len, NULL, 0) == -1)
			err(1, "sysctl");
		(void)printf("%s%.*s", prefix, len, buf);
		prefix = " ";
	}
	if (flags & RFLAG) {
		mib[0] = CTL_KERN;
		mib[1] = KERN_OSRELEASE;
		len = sizeof(buf);
		if (sysctl(mib, 2, &buf, &len, NULL, 0) == -1)
			err(1, "sysctl");
		(void)printf("%s%.*s", prefix, len, buf);
		prefix = " ";
	}
	if (flags & VFLAG) {
		mib[0] = CTL_KERN;
		mib[1] = KERN_VERSION;
		len = sizeof(buf);
		if (sysctl(mib, 2, &buf, &len, NULL, 0) == -1)
			err(1, "sysctl");
		for (p = buf, tlen = len; tlen--; ++p)
			if (*p == '\n' || *p == '\t')
				*p = ' ';
		(void)printf("%s%.*s", prefix, len, buf);
		prefix = " ";
	}
	if (flags & MFLAG) {
		mib[0] = CTL_HW;
		mib[1] = HW_MACHINE;
		len = sizeof(buf);
		if (sysctl(mib, 2, &buf, &len, NULL, 0) == -1)
			err(1, "sysctl");
		(void)printf("%s%.*s", prefix, len, buf);
		prefix = " ";
	}

#else	/* __GNO__ */

	if (uname(&name) == -1)
		err(1, "uname");

	if (flags & SFLAG) {
		printf("%s%s", prefix, name.sysname);
		prefix = " ";
	}
	if (flags & NFLAG) {
		printf("%s%s", prefix, name.nodename);
		prefix = " ";
	}
	if (flags & RFLAG) {
		printf("%s%s", prefix, name.release);
		prefix = " ";
	}
	if (flags & VFLAG) {
		printf("%s%s", prefix, name.version);
		prefix = " ";
	}
	if (flags & MFLAG) {
		printf("%s%s", prefix, name.machine);
		prefix = " ";
	}
#endif

	(void)printf("\n");
	exit (0);
}

void
usage __P((void))
{
	(void)fprintf(stderr, "usage: uname [-amnrsvV]\n");
	exit(1);
}
