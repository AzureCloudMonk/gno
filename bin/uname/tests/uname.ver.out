uname v3.0
